Resolver Designation (#91)
tfpauly/draft-pauly-adaptive-dns-privacy: draft-pauly-add-resolver-discovery.md

This document defines the following terms:

Direct Resolver:
: A DNS resolver using any transport, encrypted or unencrypted, that is provisioned directly by a local router or a VPN.

Designated Resolver:
: A DNS resolver that is designated as a responsible resolver for a given domain or zone. Designated resolvers use encrypted transports.

Companion DoH Server:
: A DNS resolver that provides connectivity over HTTPS (DoH) that is designated as 
equivalent to querying a particular Direct Resolver.

# Designated Resolvers

An encrypted DNS resolver, such as a DoH or DoT server, can be designated for use in resolving names within one or more zones. This means that clients can learn about an explicit mapping from a given domain or zone to one or more Designated Resolvers, and use that mapping to select the best resolver for a given query.

> Designating a resolver MUST rely on mutual agreement between the entity managing a zone (the Domain Owner) and the entity operating the resolver. These entities can be one and the same, or a Domain Owner can choose to designate a third-party resolver to handle its traffic. By presenting proof of this mutual agreement, the two entities assert to clients that interacting with the designated resolver will not expose information about names being resolved to an entity that would otherwise not be able to learn such information.

Reviewer: ```suggestion
Designating a resolver MUST rely on mutual agreement between the entity managing a zone (the Domain Owner) and the entity operating the resolver. These entities can be one and the same, or a Domain Owner can choose to designate a third-party resolver to handle its traffic. Proof of this mutual agreement asserts to clients that sending any query to the designated resolver exposes no more information than sending sending said query to the entity managing the corresponding zone.
```

(I had a hard time parsing the last sentence!)

Author: Incorporate, and fixed typo!

> As an example with only one entity, a company that runs many sites within "enterprise.example.com" can provide its own DoH resolver, "doh.enterprise.example.com", and designate only that resolver for all names that fall within "enterprise.example.com". This means that no other resolver would be designated for those names, and clients would only be resolving names with the same entity that would receive TLS connections.

Reviewer: ```suggestion
As an example with only one entity, a company that runs many sites within "enterprise.example.com" can provide its own DoH resolver, "doh.enterprise.example.com", and designate only that resolver for all names that fall within "enterprise.example.com". This means that no other resolver would be designated for those names, and clients would only resolve names with the same entity that would service TLS connections.
```

> As an example with only one entity, a company that runs many sites within "enterprise.example.com" can provide its own DoH resolver, "doh.enterprise.example.com", and designate only that resolver for all names that fall within "enterprise.example.com". This means that no other resolver would be designated for those names, and clients would only be resolving names with the same entity that would receive TLS connections.

Author: Not sure it need to be reiterated like that (at least at this stage)? This sentence is already pretty clear that the entities are the same, and I'm concerned that making more claims about what that sameness guarantees will be tricky to get 100% right.

As an example with several entities, the organization that runs sites within "example.org" may work with two different Content Delivery Networks (CDNs) to serve its sites. It might designate names under "example.com" to two different entities, "doh.cdn-a.net" and "doh.cdn-b.net". These are CDNs that have an existing relationship with the organization that runs "example.org", and have agreements with that organization about how data about names and users is handled.

There are several methods that can be used to designate a resolver:

- Based on SVCB DNS records issued to another resolver ({{svcb}})
- Based on information from Designated DoH Resolver that is confirmed via SVCB DNS records ({{pvd}})

> - Based on mutual confirmation of domains over HTTPS ({{pvd-mutual}})

Reviewer: ```suggestion
- Based on mutual agreement through confirmation of domains over HTTPS ({{pvd-mutual}})
```

(for consistency with the above text)

> Note that clients SHOULD NOT accept designations for effective top-level domains (eTLDs), such as ".com".

Reviewer: MUST NOT? Is there a case where a client might want to ignore this?

Author: Done

is a URI, which is the DoH URI template {{!RFC8484}}.

The following example shows a record containing a DoH URI, as returned by a query for
the HTTPSSVC variant of the SVCB record type on "foo.example.com", where the response indicates a DoH Resolver that is designated for names under "example.com".

~~~
   foo.example.com.  7200  IN HTTPSSVC 1 example.com. (
                           dohuri=https://doh.example.net/dns-query )
~~~

> If this record is DNSSEC-signed, clients can immediately create a mapping that indicates the server as a Designated Resolver for the name in the SVCB record.

Reviewer: ```suggestion
If this record is DNSSEC-signed, clients can immediately create a mapping that indicates the server (doh.example.net) as a Designated Resolver for the name in the SVCB record (foo.example.com).
```

For example, the PvD JSON for the DoH server "https://doh.example.net/dns-query" would be available at "https://doh.example.net/.well-known/pvd/dns-query".

Names that are listed in the "dnsZones" key in the JSON dictionary indicate other names that designate the resolver. For each of those domains, clients SHOULD issue an SVCB query to the DoH resolver. If this record confirms the designation and is DNSSEC-signed, clients can create a mapping to designate the resolver. In order to optimize the validation of these domains, servers MAY use HTTP Server Push to deliver the records prior to the request being made.

The key "dohTemplate" is also defined within the JSON dictionary ({{iana}}) to point back to the DoH URI Template itself. This is used for confirming the DoH server when the PvD is discovered locally or during mutual confirmation ({{pvd-mutual}}).

## Mutual Confirmation with PvD JSON {#pvd-mutual}

Designated DoH Resolvers that provide the PvD JSON described in {{pvd}} can also provide information to allow validation of zone designations without DNSSEC.

> The JSON dictionary MAY contains a key "trustedNames" that is an array of strings containing domains that can be used for mutual confirmation of resolver designation.

Reviewer 2: ```suggestion
The JSON dictionary MAY contain a key "trustedNames" that is an array of strings containing domains that can be used for mutual confirmation of resolver designation.
```

For example, the JSON dictionary retrieved at "https://doh.example.net/.well-known/pvd/dns-query" can contain the following contents:

~~~
   {
     "identifier": "doh.example.net.",
     "dohTemplate": "https://doh.example.net/dns-query",
     "dnsZones": ["example.com"],
     "trustedNames": ["example.com"]
   }
~~~

> This indicates that "example.com" should be treated as a designated domain, but that it can be validated by checking with the "example.com" server rather than using DNSSEC.

Reviewer: ```suggestion
This indicates that "example.com" should be treated as a designated domain, and that it can be validated by checking with the "example.com" server rather than using DNSSEC.
```

> The client then MUST issue a GET request for "https://example.com/.well-known/pvd" before trusting the designation. In order to trust the designation, this request must return valid JSON with the "dohTemplate" key indicating the original DoH resolver. For example, this dictionary could contain the following contents:

Reviewer: ```suggestion
Clients MUST perform this validation as follows. First, issue a GET request for "https://example.com/.well-known/pvd" before trusting the designation. In order to trust the designation, this request must return valid JSON with the "dohTemplate" key matching the original DoH resolver. For example, this dictionary could contain the following contents:
```

The previous text read to me like it forced clients to send a GET, whereas some clients might not care and might not want to do the validation. The intent with the new text is to say that this specific procedure MUST be followed if a client chooses to validate.

Reviewer: It might be helpful to give an example of a response that would not pass validation, e.g.:

```
~~~
   {
     "identifier": "example.com.",
     "dohTemplate": "https://not-the-doh-youre-looking-for.example.net/dns-query",
   }
~~~
```

Author: Added the example! Also reworded the GET section.
